### PR TITLE
Simulator: Added Physics Field

### LINKED ISSUE #882: SimpleSimulator: Create a PhysicsField abstraction
### Description of the task

See #768 for more information on the simple simualator

Create a PhysicsField class to make it easy to create a Field object in the Box2D physics world and convert the data to our own `Field` class. The Box2D object created in the Box2D world will be responsible for handling the physics and should be queried to get the latest Field data. The physics object should be set up so that the boundary of the field and the goals are physics objects that other things can collide with.

### Acceptance criteria

- [ ] Create a `PhysicsField` class
- [ ] Unit tests

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Added a `PhysicsField` class to work with the simulator.

### Testing Done

Unit tests

### Resolved Issues

resolves #882 

### Length Justification

Had to implement and test some extra Field functions which put it just over 500

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/world/field.h [hunk 1]
```diff
@@ -173,6 +180,15 @@ class Field
...
     Rectangle fieldLines() const;
 
+    /**
+     * Gets the area within the field boundary (the physical walls surrounding the Field).
+     * This is the entire area where the robots and ball can move, and is a superset of
+     * the area inside the field lines.
+     *
+     * @return The area within the field boundary as a rectangle
+     */
+    Rectangle fieldBoundary() const;
```
Comment (jonathanlew) [inline src/software/world/field.h:190]: :+1:

**Review by jonathanlew — DISMISSED** [review]
Looks good!

File: src/software/backend/simulation/BUILD [hunk 2]
```diff
@@ -43,3 +45,28 @@ cc_test(
         "@gtest//:gtest_main",
     ],
 )
+
+cc_library(
+    name = "field",
```
Comment (garethellis0) [inline src/software/backend/simulation/BUILD]: can we match the library name to the files just to be consistent with how we've done it elsewhere in the repo?

File: src/software/backend/simulation/box2d_util.h [hunk 3]
```diff
@@ -18,3 +20,5 @@
  * body or the world are null, false is returned.
  */
 bool bodyExistsInWorld(b2Body* body, std::shared_ptr<b2World> world);
+
+b2Vec2 createVec2(const Point& point);
```
Comment (garethellis0) [inline src/software/backend/simulation/box2d_util.h:31]: missing jdoc

File: src/software/backend/simulation/physics_field.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    field_boundary_body_def.type = b2_staticBody;
+    // Note that the body shape is defined relative to the body position. Setting the
+    // body position to (0, 0) makes it easy to add the shape using the standard field
+    // interface, since field coordinates are also relative to (0, 0)
+    // https://www.iforce2d.net/b2dtut/fixtures
+    field_boundary_body_def.position.Set(0, 0);
+    field_boundary_body = world->CreateBody(&field_boundary_body_def);
+
+    const unsigned int shape_loop_size = 4;
+    b2Vec2 vs[shape_loop_size]         = {
```
Comment (garethellis0) [inline src/software/backend/simulation/physics_field.cpp]: what does `vs` mean? Could we give it a more descriptive name? Same goes for `shape_loop_size`? Perhaps `field_boundary_vertices` and `num_field_boundary_vertices` (assuming I'm reading this correctly)?

File: src/software/backend/simulation/physics_field.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+{
+    enemy_goal_body_def.type = b2_staticBody;
+    // Note that the body shape is defined relative to the body position. Setting the
+    // body position to (0, 0) makes it easy to add the shape using the standard field
+    // interface, since field coordinates are also relative to (0, 0)
+    // https://www.iforce2d.net/b2dtut/fixtures
+    enemy_goal_body_def.position.Set(0, 0);
+    enemy_goal_body = world->CreateBody(&enemy_goal_body_def);
+
+    const unsigned int shape_chain_size          = 4;
```
Comment (garethellis0) [inline src/software/backend/simulation/physics_field.cpp]: suggest changing to `num_enemy_goal_vertices`?

File: src/software/backend/simulation/physics_field.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+{
+    friendly_goal_body_def.type = b2_staticBody;
+    // Note that the body shape is defined relative to the body position. Setting the
+    // body position to (0, 0) makes it easy to add the shape using the standard field
+    // interface, since field coordinates are also relative to (0, 0)
+    // https://www.iforce2d.net/b2dtut/fixtures
+    friendly_goal_body_def.position.Set(0, 0);
+    friendly_goal_body = world->CreateBody(&friendly_goal_body_def);
+
+    const unsigned int shape_chain_size             = 4;
```
Comment (garethellis0) [inline src/software/backend/simulation/physics_field.cpp]: suggesting changing to `num_friendly_goal_vertices`?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/backend/simulation/BUILD (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/BUILD]: agree. Changed this for Physics Ball as well

File: src/software/backend/simulation/box2d_util.h (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/box2d_util.h:31]: fixed

File: src/software/backend/simulation/physics_field.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_field.cpp]: Whoops, you're right. I wasn't paying attention and probably took this from example code. Fixed in the way you suggested.

File: src/software/backend/simulation/physics_field.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_field.cpp]: fixed

File: src/software/backend/simulation/physics_field.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_field.cpp]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by jonathanlew — APPROVED** [review]

**Review by jared-paul — APPROVED** [review]
LGTM! Just want to make sure we aren't forgetting that calling world->DestroyBody could have consequences/undefined behaviour

**Comment by MathewMacDougall** [comment]
> LGTM! Just want to make sure we aren't forgetting that calling world->DestroyBody could have consequences/undefined behaviour

Yup currently not planning on calling that in the middle of physics ticks, but how it's used will be more clear once I make a `PhysicsWorld` and start setting it up / tearing it down for tests, so keep on keeping an eye out for problems :+1:

**Pull request merged**